client/commands/v2/pysa_server.py

```diff
@@ -5,80 +5,116 @@
     async_server_connection as connection,
     start,
-    stop,
-    incremental,
 )
 from .persistent import (
     LSPEvent,
-    ServerState,
-    CONSECUTIVE_START_ATTEMPT_THRESHOLD,
     _read_lsp_request,
     try_initialize,
-    _start_pyre_server,
     _log_lsp_event,
-    parse_subscription_response,
-    StartSuccess,
-    StartFailure,
     _publish_diagnostics,
-    type_errors_to_diagnostics,
     InitializationExit,
     InitializationSuccess,
     InitializationFailure,
 )
-
+from api import query, connection as api_connection
+from api.connection import PyreQueryError
+from typing import List, Sequence, Dict
 
 LOG: logging.Logger = logging.getLogger(__name__)
 
 
-class PysaServerHandler(connection.BackgroundTask):
-    binary_location: str
-    server_identifier: str
-    pyre_arguments: start.Arguments
-    client_output_channel: connection.TextWriter
-    server_state: ServerState
+class PysaServer:
+    # I/O Channels
+    input_channel: connection.TextReader
+    output_channel: connection.TextWriter
+
+    # Immutable States
+    client_capabilities: lsp.ClientCapabilities
 
     def __init__(
         self,
+        input_channel: connection.TextReader,
+        output_channel: connection.TextWriter,
+        client_capabilities: lsp.ClientCapabilities,
+        pyre_arguments: start.Arguments,
         binary_location: str,
         server_identifier: str,
-        pyre_arguments: start.Arguments,
-        client_output_channel: connection.TextWriter,
-        server_state: ServerState,
     ) -> None:
+        self.input_channel = input_channel
+        self.output_channel = output_channel
+        self.client_capabilities = client_capabilities
+        self.pyre_arguments = pyre_arguments
         self.binary_location = binary_location
         self.server_identifier = server_identifier
-        self.pyre_arguments = pyre_arguments
-        self.client_output_channel = client_output_channel
-        self.server_state = server_state
+
+    def invalid_model_to_diagnostic(
+        self,
+        invalid_model: query.InvalidModel,
+    ) -> lsp.Diagnostic:
+        return lsp.Diagnostic(
+            range=lsp.Range(
+                start=lsp.Position(
+                    line=invalid_model.line - 1, character=invalid_model.column
+                ),
+                end=lsp.Position(
+                    line=invalid_model.stop_line - 1,
+                    character=invalid_model.stop_column,
+                ),
+            ),
+            message=invalid_model.full_error_message,
+            severity=lsp.DiagnosticSeverity.ERROR,
+            code=None,
+            source="Pysa",
+        )
+
+    def invalid_models_to_diagnostics(
+        self,
+        invalid_models: Sequence[query.InvalidModel],
+    ) -> Dict[Path, List[lsp.Diagnostic]]:
+        result: Dict[Path, List[lsp.Diagnostic]] = {}
+        for model in invalid_models:
+            result.setdefault(Path(model.path), []).append(
```

Comment:
As discussed on our call, you'll have to handle the case of a `None` path

Reply:
I added an if condition. Let me know if there's a more appropriate warning I should display.